proposals/text/7462-rag-evaluation.md

```diff
@@ -0,0 +1,236 @@
+```bash
+{'Reciprocal Rank': 0.448,
+ 'Single Hit': 0.5,
+ 'Multi Hit': 0.540,
+ 'Context Relevance': 0.537,
+ 'Faithfulness': 0.452,
+ 'Semantic Answer Similarity': 0.478
+ }
+ ```
+
+A detailed evaluation report that provides the scores of all available metrics for all queries or a subset of queries.
+
+```python
+def get_detailed_scores(query_ids: Union[List[str], str] = "all"):
+```
+
+Example output
+
+```bash
+data  = {
+    "query_id": ["53c3b3e6, 225f87f7, 8ac473ec, 97d284ca"],
+    "reciprocal_rank": [0.378064, 0.534964, 0.216058, 0.778642],
+    "single_hit": [1, 1, 0, 1],
+    "multi_hit": [0.706125, 0.454976, 0.445512, 0.250522],
+    "context_relevance": [0.805466, 0.410251, 0.750070, 0.361332],
+    "faithfulness": [0.135581, 0.695974, 0.749861, 0.041999],
+    "semantic_answer_similarity": [0.971241, 0.159320, 0.019722, 1],
+    "aggregated_score":
+      {
+          'Reciprocal Rank': 0.448,
+          'Single Hit': 0.5,
+          'Multi Hit': 0.540,
+          'Context Relevance': 0.537,
+          'Faithfulness': 0.452,
+          'Semantic Answer Similarity': 0.478
+      }
+}
+```
+
+### Comparative Evaluation Report
+
+A comparative summary that compares the performance of the model with another model based on the aggregated scores
+for all available metrics.
+
+```python
+def comparative_summary(self, other: "RAGPipelineEvaluation"):
+```
+
+```bash
+{
+    "model_1": {
+        'Reciprocal Rank': 0.448,
+        'Single Hit': 0.5,
+        'Multi Hit': 0.540,
+        'Context Relevance': 0.537,
+        'Faithfulness': 0.452,
+        'Semantic Answer Similarity': 0.478
+    },
+    "model_2": {
+        'Reciprocal Rank': 0.448,
+        'Single Hit': 0.5,
+        'Multi Hit': 0.540,
+        'Context Relevance': 0.537,
+        'Faithfulness': 0.452,
+        'Semantic Answer Similarity': 0.478
+    }
+}
+
+```
+
+A detailed comparative summary that compares the performance of the model with another model based on the scores of all
+available metrics for all queries.
+
+
+```python
+def detailed_comparative_summary(self, other: "RAGPipelineEvaluation"):
+```
+
+```bash
+{
+    "pipeline_1": {
+        "query_id": ["53c3b3e6, 225f87f7, 8ac473ec, 97d284ca"],
+        "reciprocal_rank": [0.378064, 0.534964, 0.216058, 0.778642],
+        "single_hit": [1, 1, 0, 1],
+        "multi_hit": [0.706125, 0.454976, 0.445512, 0.250522],
+        "context_relevance": [0.805466, 0.410251, 0.750070, 0.361332],
+        "faithfulness": [0.135581, 0.695974, 0.749861, 0.041999],
+        "semantic_answer_similarity": [0.971241, 0.159320, 0.019722, 1],
+        "aggregated_score":
+          {
+              'Reciprocal Rank': 0.448,
+              'Single Hit': 0.5,
+              'Multi Hit': 0.540,
+              'Context Relevance': 0.537,
+              'Faithfulness': 0.452,
+              'Semantic Answer Similarity': 0.478
+          }
+    },
+    "pipeline_2": {
+        "query_id": ["53c3b3e6, 225f87f7, 8ac473ec, 97d284ca"],
```

Comment:
query id or even query text and context are the same for all pipelines compared. We shouldn't store/show them redundantly.

Reply:
For example here is what I'd imagine in a tabular format
```
+----------------------------------+----------+------------------------+--------------------+--------------------+
|               questions          | contexts |          responses     | faithfulness pipe1 | faithfulness pipe2 |
+----------------------------------+----------+------------------------+--------------------+--------------------+
| Which is the most...             | The ...  | Football is            |                  1 |                  1 |
| What if...                       | A....    | The answer is...       |                  0 |                  1 |
+----------------------------------+----------+------------------------+--------------------+--------------------+
|                                  |          |                        |                0.5 |                  1 |
```

Reply:
true, that's a good point. this is just the output no need for that duplication either in the output.

Reply:
yes, this output can be improved, I will updated this in a bit.

I'm trying to make the output simple by using only json and lists, it's easier then to transform into any tabular format.